src/speculators/train/eagle3/core.py

```diff
@@ -0,0 +1,339 @@
+            self.hidden_size, self.draft_vocab_size, bias=False
+        )
+        # shape: [hidden_size, draft_vocab_size]
+        self.verifier_lm_head = torch.nn.Linear(
+            self.hidden_size, self.draft_vocab_size, bias=False
+        )
+
+        masked_lm_head_weight = verifier_weights["lm_head.weight"].to(
+            device=t2d.device, dtype=default_dtype
+        )[t2d.to(torch.bool), :]
+        if masked_lm_head_weight.shape != self.lm_head.weight.shape:
+            raise ValueError(
+                f"Masked verifier lm head data shape "
+                f"{masked_lm_head_weight.shape} does not match draft "
+                f"lm head shape {self.lm_head.weight.shape}"
+            )
+        self.lm_head.weight.data = masked_lm_head_weight.detach().clone()
+        self.verifier_lm_head.weight.data = masked_lm_head_weight.detach().clone()
+        self.verifier_lm_head.weight.requires_grad = False
+
+    @torch.compile
+    def forward(
+        self,
+        hidden_states: torch.Tensor,  # shape: [1, total_seq_len, 3 * hidden_size]
+        input_ids: torch.Tensor,  # shape: [1, total_seq_len]
+        lengths: torch.Tensor | None = None,  # shape: [batch_size]
+        loss_mask: torch.Tensor | None = None,  # shape: [1, total_seq_len]
+        position_ids: torch.Tensor | None = None,  # shape: [1, total_seq_len]
+        verifier_last_hidden_states: torch.Tensor
+        | None = None,  # shape: [1, total_seq_len, hidden_size]
+        ttt_steps: int | None = None,
+        use_off_policy_tokens: bool = False,
+        **kwargs,
+    ):
+        device = hidden_states.device
+        total_seq_len = hidden_states.shape[1]
+
+        if ttt_steps is None:
+            ttt_steps = self.ttt_steps
+        if lengths is None:
+            lengths = torch.tensor([total_seq_len], dtype=torch.long, device=device)
+        if position_ids is None:
+            position_ids = 1 + torch.arange(
+                total_seq_len, dtype=torch.long, device=device
+            ).unsqueeze(0)
+            # shape: [1, total_seq_len]
+
+        past_key_values = DynamicCache(config=self.decoder_layer_config)
+
+        combined_mask_mod = create_combined_mask_mod(lengths.to(device), total_seq_len)
+        block_mask = create_block_mask(
+            combined_mask_mod,
+            B=None,
+            H=None,
+            Q_LEN=total_seq_len,
+            KV_LEN=total_seq_len,
+            device=device,
+        )
+
+        hidden_states = self.fc(hidden_states)
+        # shape: [1, total_seq_len, hidden_size]
+
+        original_input_ids = input_ids.detach().clone()
+        return_loss = verifier_last_hidden_states is not None
+        if return_loss:
+            with torch.no_grad():
+                target_logits = self.verifier_lm_head(verifier_last_hidden_states)
+                # shape: [1, total_seq_len, draft_vocab_size]
+
+        loss = torch.tensor(0.0, device=device)
+        prev_correct = (
```

Comment:
i'm not sure i understand what prev_correct means in this context also, it seems like its only used in teh return loss branch so should this be moved with line 250?

Reply:
Yeah I can move this in there.